### PR TITLE
Feature/clothing management

### PR DESCRIPTION
**Pull request by arthur-barbosa18**
Thank you guys, for the opportunity!

Test done

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+COPY wait_db_to_start_server.py manage.py /opt/
+COPY src /opt/src
+COPY clothing_manager /opt/clothing_manager
+
+WORKDIR /opt/
+
+RUN pip install --use-feature=2020-resolver -r /opt/requirements.txt \
+    pip install --use-feature=2020-resolver -r /opt/requirements-dev.txt
+
+ENV DB_NAME=clothing_manager \
```
Comment (tiagoprn) [inline Dockerfile:16]: O que acha de usar as variáveis diretamente do ambiente (arquivo .env, vault de secrets, etc...)?

File: README.md [hunk 2]
```diff
@@ -1,69 +1,109 @@
...
+- Linux environment
+- Python 3+
+- Docker
 
----
+## Introduction
 
-## The Challenge
+The purpose of this software is to manage the complete clothing CRUD (create, read, update and delete).
+There are basicallyy three entities, color, promotional campaign and clothes. To register a new clothes in the system it is mandatory to inform name, base price, color, genre and collection. Already to register color is necessary name and hexadecimal code and to register a promotional campaign is necessary to inform name and percentage discount. For more informations access the swagger.
```
Comment (tiagoprn) [inline README.md:26]: Como o final_price não é obrigatório, o que acha de tirá-lo da representação do swagger para POST/PUT/PATCH?

File: README.md [hunk 3]
```diff
@@ -1,69 +1,109 @@
...
+#### Final Price
 
-- Create an endpoint to populate data into the model/table using a CSV file. One of the fields of the model/table must have its' value calculated based on 1 or more of the other ones.
+The final price of the clothes is generated by fields: release and promotional campaign. If that clothes is attached to a promotional campaign, then final price is base price plus promotional discount from campaign. And if that clothes is a release, then the base price is increased by 30%.
 
-- Create a frontend that has a list of all resources. It must consume the backend API endpoint that lists the resources.
+#### CSV files to bulk registration of clothes
 
-- Provide a way for us to run your application locally with all of its' requirements (python and infrastructure-wide)
+Due to some fields are float numeric, the delimiter is **semicolon (;)**.
```
Comment (tiagoprn) [inline README.md:36]: O que acha de adicionar um sample do CSV no repositório? Qual o comportamento da importação caso o arquivo contenha uma ou mais linhas com inconsistências?

File: Dockerfile (hunk 1 above)
Comment (arthur-barbosa18) [reply Dockerfile:16]: Verdade, eu poderia ter colocado as variáveis DB_NAME, DB_USER, DB_PASS, DB_SERVICE, DB_HOST e DB_PORT num .env, transferido ele para dentro do container e então lido de lá. 
Se fosse algo que necessitasse de segurança faria ainda muito mais sentido sim salvar as credenciais no vault e lê-las de lá. Concordo plenamente que são soluções melhores do que deixar hard coded.

File: README.md (hunk 2 above)
Comment (arthur-barbosa18) [reply README.md:26]: Então, concordo também. Eu acho que seria uma boa prática nem salvá-lo no banco, como é uma coluna que depende apenas de outros valores, não seria melhor toda vez que for acessá-lo exercer o cálculo para obtê-lo?

E sim, do jeito que fiz, como já nem é para o user preencher esse campo (o preenchimento é inútil), o ideal é retirá-lo da representação do swagger, comi bola nisso!

File: README.md (hunk 3 above)
Comment (arthur-barbosa18) [reply README.md:36]: Se ele conter uma ou mais linhas com inconsistências não passará pelas validações que fiz.

Agora sobre adicionar um sample de csv no repositório eu não sei se concordo com o upload de arquivos que não sejam códigos em repositório de código. Talvez fosse mais interessante fazer referência a um link onde estaria salvo esse csv de exemplo ou simular um csv de exemplo com markdown no README.

Comment (arthur-barbosa18) [reply README.md:36]: Mas, de fato, errei em não documentar isso.

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    -r requirements-dev.txt
+
+commands =
+    pytest --disable-pytest-warnings --junitxml test-reports/test_report.xml --cov-report term-missing --cov=src
+    coverage xml
+    coverage html
+    pylint -d duplicate-code src
+    #pycodestyle src #only migrations files have any problems
+
+[testenv:integration]
```
Comment (tiagoprn) [inline tox.ini:18]: Não temos muita experiência com o tox no time. Mas se entendemos bem, parece um pouco estranho criar e aplicar as migrations no momento em que você está no fluxo de CI e/ou CD.

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+commands =
+    {[testenv]commands}
+    python3 manage.py makemigrations #comment if you need run makemigrations
```
Comment (tiagoprn) [inline tox.ini:21]: Ele roda o makemigrations aqui e também no script "wait_db_to...."?

File: tox.ini (hunk 4 above)
Comment (arthur-barbosa18) [reply tox.ini:18]: Então, o tox pode simular steps de uma pipeline localmente, mas acredito não ser a ferramenta ideal para rodar num CI/CD.

E eu coloquei os comandos de migrations na flag integration de execução do tox imaginando que na primeira vez que rodassem, isso poderia ser útil.

File: src/clothes/tests.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+from .models import Clothes
+from .apps import ClothesConfig
+import pytest
+import pdb
+from django.urls import reverse
+
+
+@pytest.mark.django_db
+def test_app_name_is_expected():
```
Comment (tiagoprn) [inline src/clothes/tests.py:9]: Algum motivo especial para testar somente propriedade?

File: tox.ini (hunk 5 above)
Comment (arthur-barbosa18) [reply tox.ini:21]: o script wait_db_to... usei apenas no docker-compose, pois foi como conseguir subir o banco e a aplicação django. O depends_on do docker-compose espera o container do mysql subir, mas não espera ele aceitar conexões. Daí criei esse script para ser executado na aplicação ao invés de já executar o python manage.py runserver 0.0.0.0:8000.

File: src/clothes/tests.py (hunk 6 above)
Comment (arthur-barbosa18) [reply src/clothes/tests.py:9]: Aumentar cobertura de teste

File: src/colors/apps.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
+from django.apps import AppConfig
+
+
+class ColorsConfig(AppConfig):
```
Comment (tiagoprn) [inline src/colors/apps.py:4]: Algum motivo em especial para ter separado `colors, promotional_campaign e clothes` em apps django específicas?

File: src/colors/migrations/0001_initial.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    initial = True
+
+    dependencies = [
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='Color',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
```
Comment (tiagoprn) [inline src/colors/migrations/0001_initial.py:17]: Algum motivo em especial para não aplicar o linter nas migrations?

File: src/colors/apps.py (hunk 7 above)
Comment (arthur-barbosa18) [reply src/colors/apps.py:4]: Bom, confesso que não tenho tanta experiência em django, mas eu julguei não ser possível manter dois models, representando duas tabelas do banco, no mesmo app.

E porque pensei em deixar color e promotional campaign em tables separadas? Foi porque julguei que as duas podem aceitar n entradas, existem n campanhas promocionals e n cores diferentes. Não achei prudente colocar numa estrutura de enum, por exemplo, como fiz com collection e genre.

File: src/colors/tests.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
+from .models import Color
+from .serializers import ColorSerializer
+from .apps import ColorsConfig
+import pytest
+import pdb
+from django.urls import reverse
+
+
+@pytest.mark.django_db
```
Comment (tiagoprn) [inline src/colors/tests.py:9]: Gostamos do escopo dos testes nesse app. Algum motivo em especial para não haverem os mesmos testes no app de clothes, já que ele parece ser o principal e possui cálculos e potenciais corner cases?

File: src/colors/migrations/0001_initial.py (hunk 8 above)
Comment (arthur-barbosa18) [reply src/colors/migrations/0001_initial.py:17]: Como elas podem ser geradas automaticamente, não vi tanto valor em aplicar o pylint nas migrations.

File: src/colors/tests.py (hunk 9 above)
Comment (arthur-barbosa18) [reply src/colors/tests.py:9]: Sim, para ser super honesto, eu estava cansado, já era tarde e os testes do clothes eram bem mais sofisticados, ele tem mais campos campos que dependem de outras tabelas e uma rota diferente.

Eu achei que implementando os demais, mais simples, já ia passar o recado de que eu sei implementar testes automatizados.

File: src/colors/views.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    request_body=ColorSerializer,
+    responses={
+        status.HTTP_400_BAD_REQUEST: "Bad Request"
+    },
+)
+@api_view(['GET', 'PUT', 'DELETE', "PATCH"])
+def operation_by_index(request, foreign_key):
+    """ function called in http operations by index
+    request: contains json data
+    foreign_key: foreingh key id
```
Comment (tiagoprn) [inline src/colors/views.py:48]: Algum motivo especial para ter usado o nome "foreign key"?

File: src/promotional_campaigns/tests.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
+from .models import PromotionalCampaign
+from .serializers import PromotionalCampaignSerializer
+from .apps import PromotionalCampaignsConfig
+import pytest
+import pdb
```
Comment (tiagoprn) [inline src/promotional_campaigns/tests.py:5]: oops! :eyes:

File: src/colors/views.py (hunk 10 above)
Comment (arthur-barbosa18) [reply src/colors/views.py:48]: Ops, errei, era para ter sido primary_key

File: src/promotional_campaigns/tests.py (hunk 11 above)
Comment (arthur-barbosa18) [reply src/promotional_campaigns/tests.py:5]: então, ainda bem que tem revisão de código!

File: src/repo/commons.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
+import logging
+from rest_framework import status
+
+LOGGER = logging.getLogger(__name__)
```
Comment (tiagoprn) [inline src/repo/commons.py:4]: O que acha de usar como variável ao invés de constante (estamos supondo constante pelo uppercase no nome).

Comment (arthur-barbosa18) [reply src/repo/commons.py:4]: Qual a vantagem de de logging.getLogger(__name__) ser variável nesse caso?

E, bom, eu coloquei em uppercase, pois o padrão do python exige que atributos do módulo estejam em uppercase

File: src/repo/commons.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def treat_error_response_http(errors):
+    """ Responsible by treat erros to appropriate http response """
+    LOGGER.info("Start treat error fot http response")
+    errors_type = []
+    fields_with_error = [*errors]
+    for field in fields_with_error:
+        errors_type.append(errors[field][0].__dict__["code"])
```
Comment (tiagoprn) [inline src/repo/commons.py:13]: O que acha de extrair `errors[field][0].__dict__["code"]` em uma variável, para legibilidade?

Comment (arthur-barbosa18) [reply src/repo/commons.py:13]: É uma boa sugestão. 
Eu tenho o hábito de não fazer atribuições que não usarei mais de uma vez.

File: src/repo/commons.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+            "invalid" in errors_type or
+            "min_value" in errors_type):
+        return {"body": errors, "status": status.HTTP_422_UNPROCESSABLE_ENTITY}
+    if "required" in errors_type:
+        return {"body": errors, "status": status.HTTP_412_PRECONDITION_FAILED}
+    return {"body": errors, "status": status.HTTP_400_BAD_REQUEST}
+
+
+def create_data_to_bulk_insert(csv_content):
+    """ csv_content: pandas dataframe
```
Comment (tiagoprn) [inline src/repo/commons.py:26]: Se a intenção foi documentar o parâmetro e o retorno, pode ser uma boa padronizar o formato, ou talvez usar as type hints do python 3 para indicar os tipos - costumamos usar essa última abordagem. :wink:

File: src/clothes/tests.py (hunk 6 above)
Comment (arthur-barbosa18) [reply src/clothes/tests.py:9]: Se esse atributo (app name) estivesse sendo usada em outros lugares, esperariam que ela tivesse um determinado valor.
Aí teria mais valor testar se o app name tem o nome X.

File: src/repo/commons.py (hunk 14 above)
Comment (arthur-barbosa18) [reply src/repo/commons.py:26]: Excelente sugestão!
Confesso que nos projetos anteriores em que trabalhei não usávamos um padrão de docstring, apesar de comumente a colocarmos, normalmente tinha o proposito da função, variaveis de entrada (tipo), o que significava e o retorno.

Mas usar as type hints sem dúvida seria melhor.

File: src/repo/commons.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    return list of data to bulk insert
+    """
+    LOGGER.info("Start create data to bulk insert")
+    each_data = {}
+    bulk_data = []
+    for index in range(len(csv_content[[*csv_content][0]])):
```
Comment (tiagoprn) [inline src/repo/commons.py:32]: Não ficou muito claro o objetivo de se "montar" essa lista para o bulk insert. Consegue nos explicar?

Comment (arthur-barbosa18) [reply src/repo/commons.py:32]: Então, o insert espera um array de dicts
exemplo:
[ {name: xablau, base_price: 1, color: 1, promotional_campaign: 2, genre: M, release: True}, {}, {}... ]

Mas lendo o csv eu tenho na primeira linha o nome de cada campo (coluna), na segunda em diante os valores. Esse não é o formato adequado

File: src/repo/generic_view.py [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+    entity_name = entity_model._meta.db_table.lower()
+
+    entity_serializer = _get_entity_serializer(entity_serializer)
+    if request.method == 'GET':
+        LOGGER.info("Start get all %s", entity_name)
+        return _get_all(entity_model, entity_serializer, entity_name)
+
+    if request.method == "DELETE":
+        entity_model.objects.all().delete()
+        LOGGER.info("Start delete all %s", entity_name)
```
Comment (tiagoprn) [inline src/repo/generic_view.py:83]: Perigo! Perigo! Perigo! :rofl:

File: src/clothes/models.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+                                             null=True)
+    genre = models.CharField(null=False, max_length=100, choices=GENRE_ENUM)
+    collection = models.CharField(null=False, max_length=100,
+                                  choices=COLLECTION_ENUM)
+    release = models.BooleanField(default=False)
+
+    class Meta:
+        db_table = "Clothes"
+
+    def _str_(self):
```
Comment (rockyshimithy) [inline src/clothes/models.py:44]: Aqui não seria ? :thinking: 
```suggestion
    def __str__(self):
```

Comment (arthur-barbosa18) [reply src/clothes/models.py:44]: Verdade, erro de digitação

File: src/clothes/models.py [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    ("U", "unissex")
+)
+
+COLLECTION_ENUM = (
+    ("AW", "autumn-winter"),
+    ("SS", "sprint-summer")
+)
+
+
+RELEASE_ADJUST_PRICE = 30
```
Comment (rockyshimithy) [inline src/clothes/models.py:18]: O que acha de esse valor poder ser uma configuração?

File: src/clothes/serializers.py [hunk 19]
```diff
@@ -0,0 +1,10 @@
+from rest_framework import serializers
+from .models import Clothes
+
+
+class ClothesSerializer(serializers.ModelSerializer):
+    """ Class to define Clothes data serializer """
+    class Meta:
+        model = Clothes
+        fields = ("name", "base_price", "final_price", "color",
```
Comment (rockyshimithy) [inline src/clothes/serializers.py:9]: O que acha de informar o identificador do objeto também?

File: src/clothes/urls.py [hunk 20]
```diff
@@ -0,0 +1,13 @@
+from django.urls import path
+
+from . import views
+
+urlpatterns = [
+    path("clothes", views.create_get_delete, name="clothes"),
```
Comment (rockyshimithy) [inline src/clothes/urls.py:6]: Por que algumas rotas terminam com `/` e outras não? :thinking:

File: src/clothes/urls.py [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+    path("clothes/<int:foreign_key>/",
+         views.operation_by_index,
+         name="clothes"),
```
Comment (rockyshimithy) [inline src/clothes/urls.py:9]: Por que as rotas possuem os mesmos nomes?

File: src/clothes/views.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+        responses={
+            status.HTTP_400_BAD_REQUEST: ErrorSerializer(),
+        },
+    )
+    @api_view(["POST"])
+    def create_by_csv_file(request):
+        """ make a bulk insert to create clothes by csv upload
+        request: contains csv file instance
+        """
+        if request.method == "POST":
```
Comment (rockyshimithy) [inline src/clothes/views.py:89]: É necessário fazer esse `if` mesmo já tendo o `@api_view(["POST"])` decorado?

File: src/clothes/views.py [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+            clothes_serializer = ClothesSerializer(data=data, many=True)
+            if clothes_serializer.is_valid():
+                clothes_serializer.save()
+                LOGGER.info("Clothes by csv was created")
+                return Response(clothes_serializer.data,
+                                status=status.HTTP_201_CREATED)
+            LOGGER.error("Some error occour: %s", clothes_serializer.errors)
+            return Response(clothes_serializer.errors,
+                            status.HTTP_400_BAD_REQUEST)
+        LOGGER.error("Some error occour: %s")
```
Comment (rockyshimithy) [inline src/clothes/views.py:103]: :thinking: acho que aqui faltaria um argumento não? E imagino que talvez essas duas últimas linhas não seriam necessárias dado que sempre vai entrar aqui nesse método como post, vide o decorator, certo?

Comment (arthur-barbosa18) [reply src/clothes/views.py:103]: Não falta, ele vai criar clothes pelo csv, o body com o file .csv é acessível em request e o tipo de método também.

E sobre as duas últimas linhas serem desnecessárias acredito que você tenha razão, precisaria de testar para ter certeza.

File: src/clothes/urls.py (hunk 21 above)
Comment (arthur-barbosa18) [reply src/clothes/urls.py:9]: para elas ficarem agrupadas no swagger sob a mesma entidade

File: src/clothes/urls.py (hunk 20 above)
Comment (arthur-barbosa18) [reply src/clothes/urls.py:6]: As barras que mantive ao final da url foram por falta de atenção

File: src/clothes/models.py (hunk 18 above)
Comment (arthur-barbosa18) [reply src/clothes/models.py:18]: É uma melhoria útil

File: src/colors/migrations/0001_initial.py (hunk 8 above)
Comment (rockyshimithy) [reply src/colors/migrations/0001_initial.py:17]: E nos casos em que a migration precise ser customizada?  Exemplos: Rollback customizado.

File: src/clothes/serializers.py (hunk 19 above)
Comment (arthur-barbosa18) [reply src/clothes/serializers.py:9]: Então, não sei se é uma boa, porque o id aqui é gerado automaticamente na hora de salvar. E eu não vejo valor em validar se o body de uma atualização contém o id. Talvez seja melhor mantê-lo como read-only. O que você acha?

File: src/clothes/views.py (hunk 22 above)
Comment (arthur-barbosa18) [reply src/clothes/views.py:89]: Não, bem observado, é desnecessário.

File: src/promotional_campaigns/models.py [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+
+class PromotionalCampaign(models.Model):
+    """ Responsible by define Color table fields """
+    name = models.CharField(max_length=100, unique=True)
+    percentage_discount = models.IntegerField(
+        validators=[MinValueValidator(1), MaxValueValidator(100)])
+
+    class Meta:
+        """ Define Color table name """
+        db_table = "Promotional Campaings"
```
Comment (rockyshimithy) [inline src/promotional_campaigns/models.py:13]: Por que deixar o nome da tabela com letras maiúsculas e espaços, sendo que a sugestão da documentação é lowercase e underline?

File: src/repo/generic_view.py [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+from typing import Callable
+from django.http.response import JsonResponse
+from rest_framework.response import Response
+from rest_framework import status
+from .commons import treat_error_response_http
+
+LOGGER = logging.getLogger(__name__)
+
+
+def _get_entity_serializer(entity_serializer) -> Callable:
```
Comment (rockyshimithy) [inline src/repo/generic_view.py:11]: Por que apenas aqui usa type hint? E nesse caso do retorno, não seria `Union[Callable, None]` ?

File: src/promotional_campaigns/models.py (hunk 24 above)
Comment (arthur-barbosa18) [reply src/promotional_campaigns/models.py:13]: Falta de atenção, é muito ruim dar nomes com caracteres especiais, como espaço, para tabelas, pasta, arquivo, etc.

File: src/repo/generic_view.py [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+    """ get Entity Serializer if callable """
+    if callable(entity_serializer):
+        return entity_serializer
+    return None
```
Comment (rockyshimithy) [inline src/repo/generic_view.py:15]: Qual o propósito dessa função se ela retorna o próprio serializer, e se não for callable, quem for usar o retorno dessa função vai sofrer um side effect ?

File: src/repo/generic_view.py (hunk 25 above)
Comment (arthur-barbosa18) [reply src/repo/generic_view.py:11]: Então, usar type hint seria melhor nos casos em que estou vendo a necessidade de informar o tipo dos parâmetros. Só me falta o hábito de usá-la. 
E sim, Union[Callable, None] seria o adequado.

File: src/repo/generic_view.py (hunk 26 above)
Comment (arthur-barbosa18) [reply src/repo/generic_view.py:15]: Pylint não entende que o parâmetro entity_serializer é callable, por isso fiz essa função.

File: src/repo/generic_view.py (hunk 16 above)
Comment (rockyshimithy) [reply src/repo/generic_view.py:83]: :open_mouth:

File: requirements.txt [hunk 27]
```diff
@@ -0,0 +1,9 @@
+coreapi==2.3.3
+Django==2.2.4
+django-rest-swagger==2.2.0
+djangorestframework==3.11.0
+drf-yasg==1.17.1
+mysqlclient==2.0.1
+openapi-codec==1.3.2
+requests==2.24.0
```
Comment (rockyshimithy) [inline requirements.txt:8]: No que é usado requests? :thinking:

Comment (arthur-barbosa18) [reply requirements.txt:8]: Não é usado, é resquício de outras pocs que fiz e esqueci de remover.

File: src/colors/migrations/0001_initial.py (hunk 8 above)
Comment (arthur-barbosa18) [reply src/colors/migrations/0001_initial.py:17]: Então, o caso de uso desse teste não considerei esses casos. 
Para esse caso de uso não há rollback, a migration vai ser gerada uma vez, automaticamente, para cada máquina em que a aplicação for executada.

File: src/repo/commons.py (hunk 15 above)
Comment (mickawaii) [reply src/repo/commons.py:32]: Eu acho que tem uma forma mais fácil de montar um dicionário a partir de um pandas dataframe: https://www.laurivan.com/save-pandas-dataframe-as-django-model/

Comment (arthur-barbosa18) [reply src/repo/commons.py:32]: Eu cheguei a tentar isso, mas não deu certo e não tive tempo para estressar a solução. Mas concordo que é uma alternativa mais sofisticada do que a que implementei.
